### PR TITLE
Port AsyncSpanEndStrategy to Instrumenter API

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
This PR ports https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2530 to Instrumenter API - we need this to be able to port `opentelemetry-annotations` and `spring-boot-autoconfigure` to Instrumenter API. 

CC @HaloFour (I can't add you as a reviewer for some reason)

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (HaloFour) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:45]: Are `shouldStart`/`start` expected to ever be called through `AsyncInstrumenter`?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (laurit) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java:16]: shouldn't this be something weak

**Comment by HaloFour** [comment]
I think it looks good.

As a minor refactor of the async span strategies would you want me to follow this up with a refactor of the implementations of those strategies in the other instrumentation modules?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:6]: Does this need to be in instrumentation API? It seems to only really be needed by one instrumentation, the annotations Instrumentation, isn't it?

Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:6]: And springboot autoconfigure

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:45]: Good question -- it probably won't ever be called in this codebase. To be honest, this class is probably one that I like the least in this PR. Maybe I should have named it `AsyncInstrumenterEnd` and only keep `asyncEnd()` method

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java:16]: Actually no, not this one - these strategies shouldn't expire and get removed automatically, once installed they should be always available until unregistered (or the app shuts down)

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:6]: Yep, we need to keep it in some shared place.

> I did suggest at that time that maybe this was a little too specialized for instrumentation-api and maybe it would make sense to have a separate shared module for AOP concerns?

Hmm, I wonder about that - I fear that introducing yet another module may only increase the cognitive load of implementing an instrumentation.

@anuraaga WDYT about marking the whole package as `@UnstableApi`?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java:16]: What if the application that contains the library with async strategy gets undeployed? Won't keeping a strong ref here prevent gc of that applications class loader?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (anuraaga) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:6]: Cognitive load is precisely why we might want to find some split for it. I think async return value instrumentation is far less common than others so having a package in our instrumentation API could be too discoverable for our generally expected user. Just because a server framework is async, returning CompletableFuture, doesn't mean it needs to care about this package but someone might think so just by the word async, I would have I think.

Marking as UnstableApi seems like a good idea, can we maybe name the package annotationsasyncsupport or something similar to tie it closer to the instrumentation?

Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java:6]: > Hmm, I wonder about that - I fear that introducing yet another module may only increase the cognitive load of implementing an instrumentation.

I guess it depends on what instrumentation we expect to rely on these classes.  If it's just the AOP instrumentation (`opentelemetry-annotations-1.0` and `spring-boot-autoconfigure`) as well as the different reactive frameworks (`guava`,`rxjava2`, `rxjava3` and `reactor`, with maybe a couple more being added over time), I don't think it's too problematic to spit it into a separate module as it's use will be fairly specialized.

However, if it's expected that instrumentation for various reactive client libraries (e.g. `lettuce`, `spring-webflux-5.0`) might also want to reuse the strategies, then it probably makes more sense for them to live in `instrumentation-api`.  That said I don't think the strategies are geared towards that use, especially now that some of them are configuration driven and there's an expectation that they are registered with the reactive framework they represent.  Although nothing would stop that instrumentation module from creating its own instance of the strategy rather than bouncing through `AsyncEndStrategies`.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java:16]: > What if the application that contains the library with async strategy gets undeployed? Won't keeping a strong ref here prevent gc of that applications class loader?

Oh, that's an excellent point. `AsyncOperationEndStrategy` instances are created just to be registered in this class, they're not being referenced in any other context - a weak-keys CL->strategy map wouldn't work, since the strategy references the CL; a weak-values one wouldn't work either, because strategies could disappear even while they're still needed... Any ideas?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java]: > having a package in our instrumentation API could be too discoverable for our generally expected user. Just because a server framework is async, returning CompletableFuture, doesn't mean it needs to care about this package but someone might think so just by the word async, I would have I think.

That's very true 👍 💯 

> we maybe name the package annotationsasyncsupport or something similar to tie it closer to the instrumentation?

Done.

You've completely convinced me about the split (it'd make instrumentation-api much cleaner!), now I'm starting to think if we have any other scenarios like this one that can benefit from it 😄

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java]: The api could say that it is users responsibility to keep strategy alive e.g. strategy instances could be singletons. Or we could build an utility that allows inserting strong reference from class loader to object and use that to bind `AsyncOperationEndStrategy` instance to the class loader that defined it so that we could use weak refs without worrying that the strategy instance gets collected prematurely.

Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java]: Would that work with the ways that `AsyncSpanEndStrategy` is implemented now?  What in the different reactive modules would hold onto the root reference of the strategy?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java]: _Bikeshedding:_  `AsyncOperationEndInstrumenter`, to sync the name with the strategy interface name?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java]: I think `AsyncSpanEndStrategy` has the same problem - the non-enum strategies don't have any singleton instance and putting them in a weak reference would make them disappear.

> What in the different reactive modules would hold onto the root reference of the strategy?

In our javaagent instrumentations we can just create static singleton instances of them and that would solve the problem of weak references, but I have no idea what to do in library instrumentations, like spring-boot-autoconfigure.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncInstrumenter.java]: 👍 
I ended up with `AsyncOperationEndSupport`

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java:15]: I don't quite follow - why is this? Is it because some return types don't have one response (e.g., a reactive stream that isn't a mono)? Can we at least resolve it for the mono-like cases (especially CompletableFuture)?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndStrategies.java
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndStrategies.java:33]: Do we need this? It seems to not be symmetrical

Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndStrategies.java:33]: I believe this was also inherited from `AsyncSpanEndStrategy`.  The strategies were initially singletons so "unregistering" them by instance was sufficient but when they became configurable I wanted a way to unregister by type so that I didn't have to track strategy instances in the instrumentation, which generally has no state.

I'm not arguing for keeping that method or porting it to the new API, just providing context as to why that method was originally added.

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndStrategies.java:33]: On the other hand: if those strategies were stores as fields/constants, there would be no need for this method; we could unregister them by reference.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java:15]: > Is it because some return types don't have one response (e.g., a reactive stream that isn't a mono)?

That was one reason; the other "reason" (well, "excuse" would be more correct 😅 ) was that all those strategies pretty much operate on raw/unchecked types, before passing the response to the Instrumenter I'd have to check if it's an instance of RESPONSE type - I'd need to pass `Class<RESPONSE>` somewhere to do that, and I wanted to keep those strategies as simple as possible.
I'll work on that; multi-result reactive types should be the only thing that's not passed to the instrumenter.

Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java:15]: Having gone down the Java generics and reflection rabbit hole recently this sounds like it would be very tricky to really get right. Would the strategy itself be responsible for trying to negotiate the actual type parameters of the returned generic type?

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java:15]: I thought to make it as simple as possible, just use a `Class<RESPONSE>` instance to check type and cast, without delving into Java generics weirdness. All `Instrumenter`s in this project that would use the async support would use `Void` as the response type, so we'd never actually need to pass anything there - at least in this repo.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java]: Yeah, switching the implementation sounds like a reasonable idea. I'll try it out.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java
Comment (HaloFour) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java:15]: Maybe I'm not understanding.  Is the goal to be able to derive that `RESPONSE` is `String` and to pass a `Class<String>` to the instrumentation if the annotated method returns `CompletableFuture<String>` or `Mono<String>` or the like?  Or is this for non-AOP instrumentation?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/async/AsyncEndStrategies.java]: Done 👍

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java]: I went with the simplest possible approach:

```java
  public static <REQUEST, RESPONSE> AsyncOperationEndSupport<REQUEST, RESPONSE> create(
      Instrumenter<REQUEST, RESPONSE> syncInstrumenter,
      Class<RESPONSE> responseType,
      Class<?> asyncType)
```

You have to manually pass the expected response type. Extracting it from the passed `Instrumenter` is impossible, it will never have type parameters in the runtime.

**Comment by mateuszrzeszutek** [comment]
> As a minor refactor of the async span strategies would you want me to follow this up with a refactor of the implementations of those strategies in the other instrumentation modules?

Oh, sorry @HaloFour , I haven't noticed your comment before, it got overshadowed by all the other ones 🙇 

If you only find some time for that, yes please! That'd be really helpful.

**Comment by HaloFour** [comment]
@mateuszrzeszutek 

> If you only find some time for that, yes please! That'd be really helpful.

Can do, should I start a branch off of this one or would it be better to wait until this is approved/merged?

**Comment by mateuszrzeszutek** [comment]
> Can do, should I start a branch off of this one or would it be better to wait until this is approved/merged?

@anuraaga will you find some time to look at this PR this week? If not, then it'll probably be better for @HaloFour to branch off.

**Review by anuraaga — APPROVED** [review]

**Pull request merged**
